### PR TITLE
[Attestation] Update sgx_ql_qve_collateral_t collateral version check to include v3.1

### PR DESCRIPTION
**Pull request by RRathna**
This PR is a work in progress. Currently gathering requirements here.

Adopted changes from PR #4272 - Support for Intel PCCS API v3 by Daniel Weiße [dw@edgeless.systems](mailto:dw@edgeless.systems)

Signed-off-by: Rathna Ramesh <rathna1993@gmail.com>

### CODE REVIEW

File: host/sgx/sgxquoteprovider.c [hunk 1]
```diff
@@ -92,7 +92,7 @@ oe_result_t oe_get_sgx_quote_verification_collateral(
     }
 
     if (collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_1 &&
-        collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_3)
```
Comment (mingweishih) [inline host/sgx/sgxquoteprovider.c:95]: What happen if the version is `SGX_QL_QVE_COLLATERAL_VERSION_3` with this change?

**Review by mingweishih — COMMENTED** [review]
This looks like a breaking change. Please update the changelog as well.
It's also not cleared if there any impact on the existing users.

File: host/sgx/sgxquoteprovider.c (hunk 1 above)
Comment (yentsanglee) [reply host/sgx/sgxquoteprovider.c:95]: Don't drop support for version 3.0.

**Comment by yentsanglee** [comment]
Please check out the new definition of `sgx_ql_qve_collateral_t` in [version 1.12.1 spec](https://download.01.org/intel-sgx/sgx-dcap/1.12.1/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf#%5B%7B%22num%22%3A56%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C69%2C177%2C0%5D).

The diffs between 3.0 and 3.1 a in the comments:

```
union {
 uint32_t version; ///<‘version’ is the backward compatible legacy
 ///< representation
 struct {
 uint16_t major_version; ///< For PCS V1 and V2 APIs, the
Intel® Software Guard Extensions (Intel® SGX) Data Center Attestation Primitives: 
ECDSA Quote Library API
- 48 -
 uint16_t minor_version; ///< major_version = 1 and minor_version = 0
 } ///< the CRLs will be formatted in PEM. For
 }; ///< PCS V3 APIs, the major_version = 3 and the
 ///< minor_version can be either 0 or 1. A
 ///< minor_verion of 0 indicates the CRL’s are
 ///< formatted in Base16 encoded DER. A minor
 ///< version of 1 indicates the CRL’s are
 ///< formatted in raw binary DER.
```

**Comment by yentsanglee** [comment]
New test cases for API version 3.1 are required.

File: host/sgx/sgxquoteprovider.c (hunk 1 above)
Comment (Shivr-ctrl) [reply host/sgx/sgxquoteprovider.c:95]: These changes seem insufficient. The format of CRL is changing with 3.1. Please consider changing CRL reading code too, else failure will shift from this point to later in code where CRL is being interpreted.

Comment (qiucwang) [reply host/sgx/sgxquoteprovider.c:95]: `OE` is able to handle both PEM CRL and DER CRL, but it's not able to handle hex string CRL. For version < 3.0, CRL is a PRM, for version > 3.0 (e.g. 3.1), CRL is a DER. For version == 3.0, SGX returns a hex string that OE is not able to handle. But Azure DCAP client somehow is able to store a DER CRL for 3.0 collateral as is now so now OE is compatible with a V3 Azure DCAP collateral. Is Azure DCAP client going to retrieve a hex string CRL for 3.0 collateral in the future @Shivr-ctrl?

**Comment by yentsanglee** [comment]
@RRathna  you can get a sample version 3.1 CRL by going to 

https://sbx.api.trustedservices.intel.com/sgx/certification/v3/pckcrl?ca=processor&encoding=der

It will return a CRL in raw DER binary format.

File: host/sgx/sgxquoteprovider.c (hunk 1 above)
Comment (thomasten) [reply host/sgx/sgxquoteprovider.c:95]: Note that #4272 adds support for hex CRL

**Comment by RRathna** [comment]
bors try

File: host/sgx/sgxquoteprovider.c (hunk 1 above)
Comment (RRathna) [reply host/sgx/sgxquoteprovider.c:95]: Adopted changes from PR #4272, to support hex CRLs

**Comment by yentsanglee** [comment]
Please add test cases for the new collaterals

**Review by yentsanglee — CHANGES_REQUESTED** [review]
Add test cases

**Comment by RRathna** [comment]
PR #4420 adds these changes, with cherry-picks from Daniel weisse's branch.

**Pull request closed without merging**